### PR TITLE
cdc: do not change frozen list type in cdc log table

### LINKED ISSUE #6172: CDC: Frozen lists are serialized incorrectly in CDC log
(discovered by @aleksbykov, [here](https://github.com/scylladb/scylla/issues/6050#issuecomment-611995704))

A column of type `frozen<list<T>>` will have a corresponding column of type `frozen<map<timeuuid, T>>` in CDC log:

```cql
CREATE TABLE ks.cf_scylla_cdc_log (
    "cdc$stream_id" blob,
    "cdc$time" timeuuid,
    "cdc$batch_seq_no" int,
    "cdc$deleted_v" boolean,
    "cdc$operation" tinyint,
    "cdc$ttl" bigint,
    pk int,
    v frozen<map<timeuuid, text>>,
    PRIMARY KEY ("cdc$stream_id", "cdc$time", "cdc$batch_seq_no")
) WITH CLUSTERING ORDER BY ("cdc$time" ASC, "cdc$batch_seq_no" ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = 'CDC log for ks.cf'
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

CREATE TABLE ks.cf (
    pk int PRIMARY KEY,
    v frozen<list<text>>
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

cdc = {'postimage': 'false', 'preimage': 'true', 'enabled': 'true', 'ttl': '86400'}
```

When preparing mutation to the log table, if `transformer::transform` encounters an update of an atomic column, it just copies the value directly to the corresponding column in the log row, without any re-serialization. The function does not have special handling for frozen list case, therefore the serialized value of frozen list is copied to the frozen map column, which results in cells that cannot be deserialized as map:

```cql
cqlsh:ks> insert into cf (pk, v) values (0, ['a', 'b']);
cqlsh:ks> select v from cf_scylla_cdc_log ;
Traceback (most recent call last):
  File "/home/piodul/code/scylla-tools-java/bin/cqlsh.py", line 1053, in perform_simple_statement
    result = future.result()
  File "/home/piodul/code/scylla-tools-java/bin/../lib/cassandra-driver-internal-only-3.11.0-bb96859b.zip/cassandra-driver-3.11.0-bb96859b/cassandra/cluster.py", line 3925, in result
    raise self._final_exception
DriverException: Failed decoding result column "v" of type map<timeuuid, varchar>: bytes is not a 16-char string
```

### PR DESCRIPTION
**Pull request by piodul**
For a column of type `frozen<list<T>>` in base table, a corresponding
column of type `frozen<map<timeuuid, T>>` is created in cdc log.

Although a similar change of type takes place in case of non-frozen
lists, this is unneeded in case of frozen lists - frozen collections are
atomic, therefore there is no need for complicated type that will be
able to represent a column update that depends on its previous value
(e.g. appending elements to the end of the list).

Moreover, only cdc log table creation logic performs this type change
for frozen lists. The logic of `transformer::transform`, which is
responsible for creation of mutations to cdc log, assumes that atomic
columns will have their types unchanged in cdc log table. It simply
copies new value of the column from original mutation to the cdc log
mutation. A serialized frozen list might be copied to a field that is of
frozen map type, which may cause the field to become impossible to
deserialize.

This patch causes frozen list base table columns to have a corresponding
column in cdc log with the same type.

A test is added which asserts that the type of cdc log columns is not
changed in the case of frozen base columns.

Tests: unit(dev)
Fixes #6172

### CODE REVIEW

File: cdc/log.cc [hunk 1]
```diff
@@ -401,8 +401,11 @@ static schema_ptr create_log_schema(const schema& s, std::optional<utils::UUID>
             auto type = column.type;
             if (is_data_col) {
                 type = visit(*type, make_visitor(
-                    // lists are represented as map<timeuuid, value_type>. Otherwise we cannot express delta
+                    // non-frozen lists are represented as map<timeuuid, value_type>. Otherwise we cannot express delta
                     [] (const list_type_impl& type) -> data_type {
+                        if (type.is_atomic()) {
+                            return type.freeze();
```
Comment (haaawk) [inline cdc/log.cc:407]: You can probably just return type since it's already frozen, no?

Comment (kbr-) [reply cdc/log.cc:407]: also the `type = type->freeze()` line (416) is redundant. But nvm.

Comment (piodul) [reply cdc/log.cc:407]: @haaawk Changed `freeze()` to `shared_from_this()`.

Comment (haaawk) [reply cdc/log.cc:407]: maybe return column.type instead?

Comment (kbr-) [reply cdc/log.cc:407]: or rename `type` to `typ`, catch `&type`, and return it. (you can even `std::move` it).

Comment (haaawk) [reply cdc/log.cc:407]: Yes. That would work too.

Comment (piodul) [reply cdc/log.cc]: @haaawk @kbr- How about now? I took a slightly different approach, and removed redundant freezing by the way.

File: cdc/log.cc [hunk 2]
```diff
@@ -410,7 +410,6 @@ static schema_ptr create_log_schema(const schema& s, std::optional<utils::UUID>
                         return type.freeze();
                     }
                 ));
-                type = type->freeze();
```
Comment (haaawk) [inline cdc/log.cc:413]: @elcallio is it ok to remove this freeze? I think it should be but want to make sure it's not too tricky for me.

File: cdc/log.cc (hunk 1 above)
Comment (haaawk) [reply cdc/log.cc]: Looks great now.

File: cdc/log.cc (hunk 2 above)
Comment (kbr-) [reply cdc/log.cc:413]: `visit` returns a frozen type for each type.

**Comment by psarna** [comment]
queued as ff80b7c3e2b15a2fb948b2971997b222fa26d92e

**Comment by psarna** [comment]
Already in master - closing

**Pull request closed without merging**